File: CSLA10/CSLA.BatchQueue/BatchQueueService.vb
Imports System.Configuration
Imports System.Messaging
Imports System.Runtime.Remoting
Imports System.Runtime.Remoting.Channels
Imports System.Runtime.Remoting.Channels.Tcp
Imports System.IO
Imports System.Runtime.Serialization.Formatters.Binary

''' <summary>
''' 
''' </summary>
Namespace Server

  ''' <summary>
  ''' Implements the batch queue service.
  ''' </summary>
  ''' <remarks>
  ''' This class implements the server-side batch queue functionality.
  ''' It must be hosted within some process, typically a Windows Service.
  ''' It may also be hosted within a console application, which is
  ''' useful for testing and debugging.
  ''' </remarks>
  Public Class BatchQueueService

    Private Shared mChannel As TcpServerChannel
    Private Shared mQueue As MessageQueue
    Private Shared mMonitor As Threading.Thread
    Private Shared WithEvents mTimer As New System.Timers.Timer
    Private Shared mRunning As Boolean
    Private Shared mActiveEntries As Hashtable = Hashtable.Synchronized(New Hashtable)

    Private Shared mSync As New Threading.AutoResetEvent(False)
    Private Shared mWaitToEnd As New Threading.ManualResetEvent(False)

    ''' <summary>
    ''' Returns the name of the batch queue server.
    ''' </summary>
    Public Shared ReadOnly Property Name() As String
      Get
        Return LISTENER_NAME()
      End Get
    End Property

#Region " Start/Stop "

    ''' <summary>
    ''' Starts the batch queue.
    ''' </summary>
    ''' <remarks>
    ''' Call this as your Windows Service starts up to
    ''' start the batch queue itself. This will cause
    ''' the queue to start listening for user requests
    ''' via remoting and to the MSMQ for queued jobs.
    ''' </remarks>
    Public Shared Sub Start()

      mTimer.AutoReset = False

      ' open and/or create queue
      If MessageQueue.Exists(QUEUE_NAME) Then
        mQueue = New MessageQueue(QUEUE_NAME)

      Else
        mQueue = MessageQueue.Create(QUEUE_NAME)
      End If
      mQueue.MessageReadPropertyFilter.Extension = True

      ' start reading from queue
      mRunning = True
      mMonitor = New Threading.Thread(AddressOf MonitorQueue)
      mMonitor.Name = "MonitorQueue"
      mMonitor.Start()

      ' start remoting for Server.BatchQueue
      If mChannel Is Nothing Then
        ' set application name (virtual root name)
        RemotingConfiguration.ApplicationName = LISTENER_NAME()

        ' set up channel
        Dim properties As New Hashtable()
        properties("name") = "TcpBinary"
        properties("priority") = "2"
        properties("port") = CStr(PORT())
        Dim svFormatter As New BinaryServerFormatterSinkProvider()

        'TODO: uncomment the following line for .NET 1.1
        'svFormatter.TypeFilterLevel = Runtime.Serialization.Formatters.TypeFilterLevel.Full

        mChannel = New TcpServerChannel(properties, svFormatter)
        Channels.ChannelServices.RegisterChannel(mChannel)

        ' register our class
        RemotingConfiguration.RegisterWellKnownServiceType( _
          GetType(Server.BatchQueue), _
          "BatchQueue.rem", _
          WellKnownObjectMode.SingleCall)

      Else
        mChannel.StartListening(Nothing)
      End If

      Dim sb As New Text.StringBuilder()

      sb.Append("Batch queue processor started")
      sb.Append(vbCrLf)
      sb.AppendFormat("Name: {0}", Name)
      sb.Append(vbCrLf)
      sb.AppendFormat("Port: {0}", PORT)
      sb.Append(vbCrLf)
      sb.AppendFormat("Queue: {0}", QUEUE_NAME)
      sb.Append(vbCrLf)
      sb.AppendFormat("Max jobs: {0}", MAX_ENTRIES)
      sb.Append(vbCrLf)
      System.Diagnostics.EventLog.WriteEntry( _
        Name, sb.ToString, EventLogEntryType.Information)

    End Sub

    ''' <summary>
    ''' Stops the batch queue.
    ''' </summary>
    ''' <remarks>
    ''' <para>
    ''' Call this as your Windows Service is stopping. It
    ''' stops the batch queue, causing it to stop listening
    ''' for user requests via remoting and to stop looking for
    ''' jobs in the MSMQ queue.
    ''' </para><para>
    ''' NOTE that this method will not return until any
    ''' currently active (executing) batch jobs have completed.
    ''' </para>
    ''' </remarks>
    Public Shared Sub [Stop]()

      ' stop remoting for Server.BatchQueue
      mChannel.StopListening(Nothing)

      ' signal to stop working 
      mRunning = False
      mSync.Set()
      mMonitor.Join()
      ' close the queue
      mQueue.Close()

      If mActiveEntries.Count > 0 Then
        ' wait for work to end
        mWaitToEnd.WaitOne()
      End If

    End Sub

#End Region

#Region " Process messages "

    ' this will be running on a background thread
    Private Shared Sub MonitorQueue()

      While mRunning
        ScanQueue()
        mSync.WaitOne()
      End While

    End Sub

    ' this runs on a threadpool thread
    Private Shared Sub mTimer_Elapsed(ByVal sender As Object, _
        ByVal e As System.Timers.ElapsedEventArgs) Handles mTimer.Elapsed

      mTimer.Stop()
      mSync.Set()

    End Sub

    ' this is called by MonitorQueue
    Private Shared Sub ScanQueue()

      Dim msg As Message
      Dim holdUntil As Date
      Dim nextWake As Date = Date.MaxValue

      Dim en As MessageEnumerator = mQueue.GetMessageEnumerator
      While en.MoveNext()
        msg = en.Current
        holdUntil = CDate(Text.Encoding.ASCII.GetString(msg.Extension))
        If holdUntil <= Now Then
          If mActiveEntries.Count < MAX_ENTRIES() Then
            ProcessEntry(mQueue.ReceiveById(msg.Id))

          Else
            ' the queue is busy, go to sleep
            Exit Sub
          End If

        ElseIf holdUntil < nextWake Then
          ' find the minimum holduntil value
          nextWake = holdUntil
        End If
      End While

      If nextWake < Date.MaxValue AndAlso nextWake > Now Then
        ' we have at least one entry holding, so set the
        ' timer to wake us when it should be run
        mTimer.Interval = nextWake.Subtract(Now).TotalMilliseconds
        mTimer.Start()
      End If

    End Sub

    Private Shared Sub ProcessEntry(ByVal msg As Message)

      ' get entry from queue
      Dim entry As BatchEntry
      Dim formatter As New BinaryFormatter()
      entry = CType(formatter.Deserialize(msg.BodyStream), BatchEntry)

      ' make active
      entry.Info.SetStatus(BatchEntryStatus.Active)
      mActiveEntries.Add(entry.Info.ID, entry.Info)

      ' start processing entry on background thread
      Threading.ThreadPool.QueueUserWorkItem(AddressOf entry.Execute)

    End Sub

    ' called by BatchEntry when it is done processing so
    ' we know that it is complete and we can start another
    ' job if needed
    Friend Shared Sub Deactivate(ByVal entry As BatchEntry)

      mActiveEntries.Remove(entry.Info.ID)
      mSync.Set()
      If Not mRunning AndAlso mActiveEntries.Count = 0 Then
        ' indicate that there are no active workers
        mWaitToEnd.Set()
      End If

    End Sub

#End Region

#Region " Enqueue/Dequeue/LoadEntries "

    Friend Shared Sub Enqueue(ByVal Entry As BatchEntry)

      Dim msg As New Message()
      Dim f As New BinaryFormatter()

      With msg
        .Label = Entry.ToString
        .Priority = Entry.Info.Priority
        .Extension = Text.Encoding.ASCII.GetBytes(CStr(Entry.Info.HoldUntil))
        Entry.Info.SetMessageID(.Id)
        f.Serialize(.BodyStream, Entry)
      End With

      mQueue.Send(msg)

      mSync.Set()

    End Sub

    Friend Shared Sub Dequeue(ByVal Entry As BatchEntryInfo)

      Dim label As String
      Dim msg As Message
      Dim msgID As String

      label = Entry.ToString
      Dim en As MessageEnumerator = mQueue.GetMessageEnumerator
      mQueue.MessageReadPropertyFilter.Label = True

      While en.MoveNext()
        If en.Current.Label = label Then
          ' we found a match
          msgID = en.Current.Id
          Exit While
        End If
      End While

      If Len(msgID) > 0 Then
        mQueue.ReceiveById(msgID)
      End If

    End Sub

    Friend Shared Sub LoadEntries(ByVal List As BatchEntries)

      ' load our list of BatchEntry objects
      Dim msgs() As Message
      Dim msg As Message
      Dim formatter As New BinaryFormatter()
      Dim de As DictionaryEntry
      Dim entry As Server.BatchEntry

      ' get all active entries
      SyncLock mActiveEntries.SyncRoot
        For Each de In Server.BatchQueueService.mActiveEntries
          List.Add(CType(de.Value, BatchEntryInfo))
        Next
      End SyncLock

      ' get all queued entries
      msgs = mQueue.GetAllMessages
      For Each msg In msgs
        entry = CType(formatter.Deserialize(msg.BodyStream), Server.BatchEntry)
        entry.Info.SetMessageID(msg.Id)
        List.Add(entry.Info)
      Next

    End Sub

#End Region

#Region " Utility functions "

    Private Shared Function QUEUE_NAME() As String
      Return ".\private$\" & ConfigurationSettings.AppSettings("QueueName")
    End Function

    Private Shared Function LISTENER_NAME() As String
      Return ConfigurationSettings.AppSettings("ListenerName")
    End Function

    Private Shared Function PORT() As Integer
      Return CInt(ConfigurationSettings.AppSettings("ListenerPort"))
    End Function

    Private Shared Function MAX_ENTRIES() As Integer
      Return CInt(ConfigurationSettings.AppSettings("MaxActiveEntries"))
    End Function

#End Region

  End Class

End Namespace

File: cslacs10/CSLA.BatchQueue/BatchQueueService.cs
using System;
using System.Collections;
using System.Threading;
using System.Messaging;
using System.Runtime.Remoting;
using System.Runtime.Remoting.Channels;
using System.Runtime.Remoting.Channels.Tcp;
using System.IO;
using System.Runtime.Serialization.Formatters.Binary;
using CSLA;

namespace CSLA.BatchQueue.Server
{
  /// <summary>
  /// Implements the batch queue service.
  /// </summary>
  /// <remarks>
  /// This class implements the server-side batch queue functionality.
  /// It must be hosted within some process, typically a Windows Service.
  /// It may also be hosted within a console application, which is
  /// useful for testing and debugging.
  /// </remarks>
  public class BatchQueueService
  {
    static TcpServerChannel _channel;
    static MessageQueue _queue;
    static Thread _monitor;
    static System.Timers.Timer _timer;
    static volatile bool _running;
    static Hashtable _activeEntries = Hashtable.Synchronized(new Hashtable());

    static AutoResetEvent _sync = new AutoResetEvent(false);
    static ManualResetEvent _waitToEnd = new ManualResetEvent(false);

    static BatchQueueService()
    {
      _timer = new System.Timers.Timer();
      _timer.Elapsed += new System.Timers.ElapsedEventHandler(Timer_Elapsed);
    }

    /// <summary>
    /// Returns the name of the batch queue server.
    /// </summary>
    public static string Name
    {
      get
      {
        return LISTENER_NAME;
      }
    }

    #region Start/Stop

    /// <summary>
    /// Starts the batch queue.
    /// </summary>
    /// <remarks>
    /// Call this as your Windows Service starts up to
    /// start the batch queue itself. This will cause
    /// the queue to start listening for user requests
    /// via remoting and to the MSMQ for queued jobs.
    /// </remarks>
    public static void Start()
    {
      _timer.AutoReset = false;

      // open and/or create queue
      if(MessageQueue.Exists(QUEUE_NAME))
        _queue = new MessageQueue(QUEUE_NAME);
      else
        _queue = MessageQueue.Create(QUEUE_NAME);
      _queue.MessageReadPropertyFilter.Extension = true;

      // start reading from queue
      _running = true;
      _monitor = new Thread(new ThreadStart(MonitorQueue));
      _monitor.Name = "MonitorQueue";
      _monitor.Start();

      // start remoting for Server.BatchQueue
      if(_channel == null)
      {
        // set application name (virtual root name)
        RemotingConfiguration.ApplicationName = LISTENER_NAME;

        // set up channel
        Hashtable properties = new Hashtable();
        properties["name"] = "TcpBinary";
        properties["port"] = PORT.ToString();
        BinaryServerFormatterSinkProvider svFormatter = 
          new BinaryServerFormatterSinkProvider();

        //TODO: uncomment the following line for .NET 1.1
        //svFormatter.TypeFilterLevel = 
        //  System.Runtime.Serialization.Formatters.TypeFilterLevel.Full;

        _channel = new TcpServerChannel(properties, svFormatter);
        ChannelServices.RegisterChannel(_channel);

        // register our class
        RemotingConfiguration.RegisterWellKnownServiceType( 
          typeof(Server.BatchQueue), "BatchQueue.rem", 
          WellKnownObjectMode.SingleCall);
      }
      else
        _channel.StartListening(null);

      System.Text.StringBuilder sb = new System.Text.StringBuilder();

      sb.AppendFormat("Batch queue processor started\n");
      sb.AppendFormat("Name: {0}\n", Name);
      sb.AppendFormat("Port: {0}\n", PORT);
      sb.AppendFormat("Queue: {0}\n", QUEUE_NAME);
      sb.AppendFormat("Max jobs: {0}\n", MAX_ENTRIES);
      System.Diagnostics.EventLog.WriteEntry(Name, sb.ToString(), 
        System.Diagnostics.EventLogEntryType.Information);
    }

    /// <summary>
    /// Stops the batch queue.
    /// </summary>
    /// <remarks>
    /// <para>
    /// Call this as your Windows Service is stopping. It
    /// stops the batch queue, causing it to stop listening
    /// for user requests via remoting and to stop looking for
    /// jobs in the MSMQ queue.
    /// </para><para>
    /// NOTE that this method will not return until any
    /// currently active (executing) batch jobs have completed.
    /// </para>
    /// </remarks>
    public static void Stop()
    {
      // stop remoting for Server.BatchQueue
      _channel.StopListening(null);

      // signal to stop working 
      _running = false;
      _sync.Set();
      _monitor.Join();
      // close the queue
      _queue.Close();

      if(_activeEntries.Count > 0)
      {
        // wait for work to end
        _waitToEnd.WaitOne();
      }
    }

    #endregion

    #region Process messages

    // this will be running on a background thread
    static void MonitorQueue()
    {
      while(_running)
      {
        ScanQueue();
        _sync.WaitOne();
      }
    }

    // this runs on a threadpool thread
    static void Timer_Elapsed(object sender, System.Timers.ElapsedEventArgs e)
    {
      _timer.Stop();
      _sync.Set();
    }

    // this is called by MonitorQueue
    static void ScanQueue()
    {
      Message msg;
      DateTime holdUntil;
      DateTime nextWake = DateTime.MaxValue;

      MessageEnumerator en = _queue.GetMessageEnumerator();
      while(en.MoveNext())
      {
        msg = en.Current;
        holdUntil = 
          DateTime.Parse(System.Text.Encoding.ASCII.GetString(msg.Extension));
        if(holdUntil <= DateTime.Now)
        {
          if(_activeEntries.Count < MAX_ENTRIES)
          {
            ProcessEntry(_queue.ReceiveById(msg.Id));
          }
          else
          {
            // the queue is busy, go to sleep
            return;
          }
        }
        else
        {
          if(holdUntil < nextWake)
          {
            // find the minimum holduntil value
            nextWake = holdUntil;
          }
        }
      }

      if(nextWake < DateTime.MaxValue && nextWake > DateTime.Now)
      {
        // we have at least one entry holding, so set the
        // timer to wake us when it should be run
        _timer.Interval = nextWake.Subtract(DateTime.Now).TotalMilliseconds;
        _timer.Start();
      }
    }

    static void ProcessEntry(Message msg)
    {
      // get entry from queue
      BatchEntry entry;
      BinaryFormatter formatter = new BinaryFormatter();
      entry = (BatchEntry)formatter.Deserialize(msg.BodyStream);

      // make active
      entry.Info.SetStatus(BatchEntryStatus.Active);
      _activeEntries.Add(entry.Info.ID, entry.Info);

      // start processing entry on background thread
      ThreadPool.QueueUserWorkItem(new WaitCallback(entry.Execute));
    }

    // called by BatchEntry when it is done processing so
    // we know that it is complete and we can start another
    // job if needed
    internal static void Deactivate(BatchEntry entry)
    {
      _activeEntries.Remove(entry.Info.ID);
      _sync.Set();
      if(!_running && _activeEntries.Count == 0)
      {
        // indicate that there are no active workers
        _waitToEnd.Set();
      }
    }

    #endregion

    #region Enqueue/Dequeue/LoadEntries

    internal static void Enqueue(BatchEntry entry)
    {
      Message msg = new Message();
      BinaryFormatter f = new BinaryFormatter();

      msg.Label = entry.ToString();
      msg.Priority = entry.Info.Priority;
      msg.Extension = 
        System.Text.Encoding.ASCII.GetBytes(
        entry.Info.HoldUntil.ToString());
      entry.Info.SetMessageID(msg.Id);
      f.Serialize(msg.BodyStream, entry);

        _queue.Send(msg);

      _sync.Set();
    }

    internal static void Dequeue(BatchEntryInfo entry)
    {
      string msgID;

      string label = entry.ToString();
      MessageEnumerator en = _queue.GetMessageEnumerator();
      _queue.MessageReadPropertyFilter.Label = true;

      while(en.MoveNext())
      {
        if(en.Current.Label == label)
        {
          // we found a match
          msgID = en.Current.Id;
        }
      }

      if(msgID != null)
        _queue.ReceiveById(msgID);
    }

    internal static void LoadEntries(BatchEntries list)
    {
      // load our list of BatchEntry objects
      BinaryFormatter formatter = new BinaryFormatter();
      Server.BatchEntry entry;

      // get all active entries
      lock(_activeEntries.SyncRoot)
      {
        foreach(DictionaryEntry de in _activeEntries)
        {
          list.Add((BatchEntryInfo)de.Value);
        }
      }

      // get all queued entries
      Message [] msgs = _queue.GetAllMessages();
      foreach(Message msg in msgs)
      {
        entry = (Server.BatchEntry)formatter.Deserialize(msg.BodyStream);
        entry.Info.SetMessageID(msg.Id);
        list.Add(entry.Info);
      }
    }

    #endregion

    #region Utility functions

    static string QUEUE_NAME
    {
      get
      {
        return @".\private$\" + ConfigurationSettings.AppSettings["QueueName"];
      }
    }

    static string LISTENER_NAME
    {
      get
      {
        return ConfigurationSettings.AppSettings["ListenerName"];
      }
    }

    static int PORT
    {
      get
      {
        return Convert.ToInt32(ConfigurationSettings.AppSettings["ListenerPort"]);
      }
    }

    static int MAX_ENTRIES
    {
      get
      {
        return Convert.ToInt32(
          ConfigurationSettings.AppSettings["MaxActiveEntries"]);
      }
    }

    #endregion


  }
}
